tests/test_rotations.py

```diff
@@ -0,0 +1,110 @@
+import numpy as np
+import pytest
+from numpy.typing import NDArray
+from scipy.spatial.transform import Rotation
+
+from timemachine.lib import custom_ops
+
+
+def hamiliton_product(q1: NDArray, q2: NDArray) -> NDArray:
+    """
+    Compute the product of two quaternions, called a hamilton product.
+
+    References
+    ----------
+    https://en.wikipedia.org/wiki/Quaternion#Hamilton_product
+    """
+    w1, x1, y1, z1 = q1[0], q1[1], q1[2], q1[3]
+    w2, x2, y2, z2 = q2[0], q2[1], q2[2], q2[3]
+    output = np.empty(4)
+    output[0] = (w1 * w2) - (x1 * x2) - (y1 * y2) - (z1 * z2)
+    output[1] = (w1 * x2) + (x1 * w2) + (y1 * z2) - (z1 * y2)
+    output[2] = (w1 * y2) - (x1 * z2) + (y1 * w2) + (z1 * x2)
+    output[3] = (w1 * z2) + (x1 * y2) - (y1 * x2) + (z1 * w2)
+    return output
+
+
+def rotate_vector_by_quaternions(coords: NDArray, quaternion: NDArray) -> NDArray:
+    """
+    Reference implementation of rotating coordinates using a quaternion
+
+    References
+    ----------
+    https://math.stackexchange.com/a/535223
+    """
+    assert coords.shape == (3,)
+    assert quaternion.shape[0] == 4
+    expanded_coords = np.concatenate([np.zeros(1), coords])
+    quaternion_conjugate = quaternion.copy()
+    # Compute quaternion conjugate by taking the conjugate of all but w
+    quaternion_conjugate[1:] *= -1
+    ham_coords = hamiliton_product(quaternion, expanded_coords)
+    return hamiliton_product(ham_coords, quaternion_conjugate)[1:]  # Truncate off w
+
+
+def convert_quaternion_for_scipy(quat: NDArray) -> NDArray:
+    """Scipy has the convention of (x, y, z, w) which is different than the wikipedia definition, swap ordering to verify using scipy.
+
+    References
+    ----------
+    https://docs.scipy.org/doc/scipy/reference/generated/scipy.spatial.transform.Rotation.from_quat.html
+    """
+    return np.append(quat[1:], [quat[0]])
+
+
+@pytest.mark.nogpu
+@pytest.mark.parametrize("seed", list(range(10)))
+@pytest.mark.parametrize("size", [1000])
+def test_reference_rotation_by_quaternion(seed, size):
+    rng = np.random.default_rng(seed)
+
+    all_coords = rng.normal(size=(size, 3)) * 100.0
+
+    quaternions = rng.normal(size=(size, 4))
+    # Normalize the quaternions to a vector quaternion, implicitly performed by Rotation
```

Comment:
nit: should this say
```suggestion
    # Normalize the quaternions to unit quaternions, representing a rotations
```
?